agent/stats/reporter/reporter.go

```diff
@@ -0,0 +1,140 @@
+package reporter
+
+import (
+	"context"
+	"errors"
+	"io"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/engine"
+	"github.com/aws/amazon-ecs-agent/agent/version"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/doctor"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/eventstream"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	tcshandler "github.com/aws/amazon-ecs-agent/ecs-agent/tcs/handler"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/tcs/model/ecstcs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils/retry"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/cihub/seelog"
+)
+
+const (
+	// The maximum time to wait between heartbeats without disconnecting
+	defaultHeartbeatTimeout = 1 * time.Minute
+	defaultHeartbeatJitter  = 1 * time.Minute
+	// Default websocket client disconnection timeout initiated by agent
+	defaultDisconnectionTimeout = 15 * time.Minute
+	defaultDisconnectionJitter  = 30 * time.Minute
+)
+
+type DockerTelemetrySession struct {
+	s                    tcshandler.TelemetrySession
+	ecsClient            api.ECSClient
+	containerInstanceArn string
+}
+
+func (session *DockerTelemetrySession) Start(ctx context.Context) error {
+	backoff := retry.NewExponentialBackoff(time.Second, 1*time.Minute, 0.2, 2)
+	for {
+		endpoint, tcsError := discoverPollEndpoint(session.containerInstanceArn, session.ecsClient)
+		if tcsError == nil {
+			tcsError = session.s.StartTelemetrySession(ctx, endpoint)
+		}
+		switch tcsError {
+		case context.Canceled, context.DeadlineExceeded:
+			return tcsError
+		case io.EOF, nil:
+			logger.Info("TCS Websocket connection closed for a valid reason")
+			backoff.Reset()
+		default:
+			seelog.Errorf("Error: lost websocket connection with ECS Telemetry service (TCS): %v", tcsError)
+			time.Sleep(backoff.Duration())
+		}
+	}
+}
+
+func NewDockerTelemetrySession(
```

Comment:
Nit: Can we put NewDockerTelemetrySession before Start? I find it more aligned, not sure why. 🤭

Reply:
Sure